site/src/xServices/workspaceSchedule/workspaceScheduleXService.ts

```diff
@@ -50,160 +53,229 @@ export type WorkspaceScheduleEvent =
+      tsTypes: {} as import("./workspaceScheduleXService.typegen").Typegen0,
+      schema: {
+        context: {} as WorkspaceScheduleContext,
+        events: {} as WorkspaceScheduleEvent,
+        services: {} as {
+          getWorkspace: {
+            data: TypesGen.Workspace
+          }
+          getTemplate: {
+            data: TypesGen.Template
+          }
+        },
       },
-    },
-    initial: "idle",
-    on: {
-      GET_WORKSPACE: "gettingWorkspace",
-    },
-    states: {
-      idle: {
-        tags: "loading",
+      initial: "idle",
+      on: {
+        GET_WORKSPACE: "gettingWorkspace",
       },
-      gettingWorkspace: {
-        entry: ["clearGetWorkspaceError", "clearContext"],
-        invoke: {
-          src: "getWorkspace",
-          id: "getWorkspace",
-          onDone: {
-            target: "gettingPermissions",
-            actions: ["assignWorkspace"],
+      states: {
+        idle: {
+          tags: "loading",
+        },
+        gettingWorkspace: {
+          entry: ["clearGetWorkspaceError", "clearContext"],
+          invoke: {
+            src: "getWorkspace",
+            id: "getWorkspace",
+            onDone: {
+              target: "gettingPermissions",
+              actions: ["assignWorkspace"],
+            },
+            onError: {
+              target: "error",
+              actions: ["assignGetWorkspaceError"],
+            },
           },
-          onError: {
-            target: "error",
-            actions: ["assignGetWorkspaceError"],
+          tags: "loading",
+        },
+        gettingPermissions: {
+          entry: "clearGetPermissionsError",
+          invoke: {
+            src: "checkPermissions",
+            id: "checkPermissions",
+            onDone: [
+              {
+                actions: ["assignPermissions"],
+                target: "gettingTemplate",
+              },
+            ],
+            onError: [
+              {
+                actions: "assignGetPermissionsError",
+                target: "error",
+              },
+            ],
           },
+          tags: "loading",
         },
-        tags: "loading",
-      },
-      gettingPermissions: {
-        entry: "clearGetPermissionsError",
-        invoke: {
-          src: "checkPermissions",
-          id: "checkPermissions",
-          onDone: [
-            {
-              actions: ["assignPermissions"],
+        gettingTemplate: {
+          entry: "clearGetTemplateError",
+          invoke: {
+            src: "getTemplate",
+            id: "getTemplate",
+            onDone: {
               target: "presentForm",
+              actions: ["assignTemplate"],
             },
-          ],
-          onError: [
-            {
-              actions: "assignGetPermissionsError",
+            onError: {
               target: "error",
+              actions: ["assignGetTemplateError"],
             },
-          ],
+          },
+          tags: "loading",
```

Comment:
Maybe, we could condense the `gettingPermissions`, `gettingWorkspace`, and `gettingTemplate` in a single state called `loading` + a `load` service that returns all this info. I'm thinking we are over-creating states making the machine more complex to understand without a clear value. By over-creating, I mean states that are not reflected in the UI/UX.

Reply:
Yeah, on one hand having separate states is helpful for debugging, but on the other hand it's extra states that aren't strictly necessary. We could change the way we handle errors to save whichever error we get going through the whole loading sequence. I'm not going to hold up this PR over it but it's worth considering in the future.